File: src/formatters.ts
import type * as srk from '@algoux/standard-ranklist';

/**
 * Convert an srk time duration to the requested unit.
 *
 * @param time - Source duration tuple, such as `[90, 's']`.
 * @param targetUnit - Unit to convert the duration to.
 * @param fmt - Optional formatter used for converted non-millisecond values, commonly `Math.floor`, `Math.ceil`, or `Math.round`.
 * @returns Converted duration value.
 * @throws If the source value is negative/non-finite, or the source/target time unit is unsupported.
 */
export function formatTimeDuration(
  time: srk.TimeDuration,
  targetUnit: srk.TimeUnit = 'ms',
  fmt: (num: number) => number = (num) => num,
) {
  if (!Number.isFinite(time[0]) || time[0] < 0) {
    throw new Error(`Invalid source time value ${time[0]}`);
  }
  let ms = -1;
  switch (time[1]) {
    case 'ms':
      ms = time[0];
      break;
    case 's':
      ms = time[0] * 1000;
      break;
    case 'min':
      ms = time[0] * 1000 * 60;
      break;
    case 'h':
      ms = time[0] * 1000 * 60 * 60;
      break;
    case 'd':
      ms = time[0] * 1000 * 60 * 60 * 24;
      break;
    default:
      throw new Error(`Invalid source time unit ${time[1]}`);
  }
  switch (targetUnit) {
    case 'ms':
      return ms;
    case 's':
      return fmt(ms / 1000);
    case 'min':
      return fmt(ms / 1000 / 60);
    case 'h':
      return fmt(ms / 1000 / 60 / 60);
    case 'd':
      return fmt(ms / 1000 / 60 / 60 / 24);
    default:
      throw new Error(`Invalid target time unit ${targetUnit}`);
  }
}

/**
 * Left-pad a number with zeroes until it reaches the requested display width.
 *
 * @param num - Number to format.
 * @param size - Minimum string width.
 * @returns Zero-filled number string, or the original number string when it is already long enough.
 */
export function preZeroFill(num: number, size: number): string {
  if (num >= Math.pow(10, size)) {
    return num.toString();
  } else {
    let str = Array(size + 1).join('0') + num;
    return str.slice(str.length - size);
  }
}

/**
 * Format elapsed seconds as a ranklist time string.
 *
 * @param second - Elapsed seconds.
 * @param options - Formatting options.
 * @param options.fillHour - Pad the hour field to two digits.
 * @param options.showDay - Show a leading day count when the duration is at least one day.
 * @returns Time string in `H:mm:ss`/`HH:mm:ss` form, optionally prefixed with `nD `; returns `--` for negative input.
 */
export function secToTimeStr(second: number, options: { fillHour?: boolean; showDay?: boolean } = {}): string {
  if (second < 0) {
    return '--';
  }
  let sec = second;
  let d = 0;
  const { fillHour = false, showDay = false } = options;
  if (showDay) {
    d = Math.floor(sec / 86400);
    sec %= 86400;
  }
  let h = Math.floor(sec / 3600);
  sec %= 3600;
  let m = Math.floor(sec / 60);
  sec %= 60;
  let s = Math.floor(sec);
  let dayStr = '';
  if (showDay && d >= 1) {
    dayStr = d + 'D ';
  }
  return dayStr + (fillHour ? preZeroFill(h, 2) : `${h}`) + ':' + preZeroFill(m, 2) + ':' + preZeroFill(s, 2);
}

/**
 * Convert a zero-based numeric problem index to an alphabetic problem alias.
 *
 * @param number - Numeric index, or a numeric string.
 * @returns Alphabetic alias such as `A`, `Z`, `AA`, or `AC`.
 * @example
 * numberToAlphabet(0) // 'A'
 * numberToAlphabet(25) // 'Z'
 * numberToAlphabet(26) // 'AA'
 * numberToAlphabet(28) // 'AC'
 */
export function numberToAlphabet(number: number | string): string {
  let n = ~~number;
  const radix = 26;
  let cnt = 1;
  let p = radix;
  while (n >= p) {
    n -= p;
    cnt++;
    p *= radix;
  }
  let res = [];
  for (; cnt > 0; cnt--) {
    res.push(String.fromCharCode((n % radix) + 65));
    n = Math.trunc(n / radix);
  }
  return res.reverse().join('');
}

/**
 * Convert an alphabetic problem alias to a zero-based numeric index.
 *
 * @param alphabet - Alphabetic alias.
 * @returns Zero-based index, or `-1` for empty/non-string input.
 * @example
 * alphabetToNumber('A') // 0
 * alphabetToNumber('Z') // 25
 * alphabetToNumber('AA') // 26
 * alphabetToNumber('AC') // 28
 */
export function alphabetToNumber(alphabet: string): number {
  if (typeof alphabet !== 'string' || !alphabet.length) {
    return -1;
  }
  const chars = `${alphabet}`.toUpperCase().split('').reverse();
  const radix = 26;
  let p = 1;
  let res = -1;
  chars.forEach((ch) => {
    res += (ch.charCodeAt(0) - 65) * p + p;
    p *= radix;
  });
  return res;
}

File: go/formatters.go
package srkutils

import (
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
)

func numberValue(value any) float64 {
	switch v := value.(type) {
	case int:
		return float64(v)
	case int64:
		return float64(v)
	case float64:
		return v
	case float32:
		return float64(v)
	case json.Number:
		f, _ := v.Float64()
		return f
	default:
		f, _ := strconv.ParseFloat(fmt.Sprint(v), 64)
		return f
	}
}

func stringValue(value any) string {
	if value == nil {
		return ""
	}
	return fmt.Sprint(value)
}

func asSlice(value any) []any {
	switch v := value.(type) {
	case []any:
		return v
	case TimeDuration:
		return []any{v.Value, string(v.Unit)}
	case []float64:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = item
		}
		return result
	case []int:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = item
		}
		return result
	case []int64:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = item
		}
		return result
	case []map[string]any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = item
		}
		return result
	case []string:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = item
		}
		return result
	case nil:
		return []any{}
	default:
		return []any{}
	}
}

func asMap(value any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	if typed, ok := value.(map[string]any); ok {
		return typed
	}
	return map[string]any{}
}

func FormatTimeDurationChecked(time any, targetUnit string, fmtFn func(float64) float64) (float64, error) {
	raw := asSlice(time)
	if len(raw) < 2 {
		return 0, fmt.Errorf("invalid source time duration")
	}
	value := numberValue(raw[0])
	unit := stringValue(raw[1])
	if math.IsNaN(value) || math.IsInf(value, 0) || value < 0 {
		return 0, fmt.Errorf("invalid source time value %v", raw[0])
	}
	ms := -1.0
	switch unit {
	case "ms":
		ms = value
	case "s":
		ms = value * 1000
	case "min":
		ms = value * 1000 * 60
	case "h":
		ms = value * 1000 * 60 * 60
	case "d":
		ms = value * 1000 * 60 * 60 * 24
	default:
		return 0, fmt.Errorf("invalid source time unit %s", unit)
	}
	if fmtFn == nil {
		fmtFn = func(number float64) float64 { return number }
	}
	switch targetUnit {
	case "", "ms":
		return ms, nil
	case "s":
		return fmtFn(ms / 1000), nil
	case "min":
		return fmtFn(ms / 1000 / 60), nil
	case "h":
		return fmtFn(ms / 1000 / 60 / 60), nil
	case "d":
		return fmtFn(ms / 1000 / 60 / 60 / 24), nil
	default:
		return 0, fmt.Errorf("invalid target time unit %s", targetUnit)
	}
}

func FormatTimeDuration(time any, targetUnit string, fmtFn func(float64) float64) float64 {
	value, err := FormatTimeDurationChecked(time, targetUnit, fmtFn)
	if err != nil {
		panic(err)
	}
	return value
}

func PreZeroFill(num int, size int) string {
	if float64(num) >= math.Pow(10, float64(size)) {
		return strconv.Itoa(num)
	}
	text := strings.Repeat("0", size) + strconv.Itoa(num)
	return text[len(text)-size:]
}

type SecToTimeStrOptions struct {
	FillHour bool
	ShowDay  bool
}

func SecToTimeStr(second float64, options SecToTimeStrOptions) string {
	if second < 0 {
		return "--"
	}
	sec := second
	days := 0
	if options.ShowDay {
		days = int(math.Floor(sec / 86400))
		sec = math.Mod(sec, 86400)
	}
	hours := int(math.Floor(sec / 3600))
	sec = math.Mod(sec, 3600)
	minutes := int(math.Floor(sec / 60))
	sec = math.Mod(sec, 60)
	seconds := int(math.Floor(sec))
	dayText := ""
	if options.ShowDay && days >= 1 {
		dayText = fmt.Sprintf("%dD ", days)
	}
	hourText := strconv.Itoa(hours)
	if options.FillHour {
		hourText = PreZeroFill(hours, 2)
	}
	return fmt.Sprintf("%s%s:%s:%s", dayText, hourText, PreZeroFill(minutes, 2), PreZeroFill(seconds, 2))
}

func NumberToAlphabet(number any) string {
	n := int(math.Trunc(numberValue(number)))
	radix := 26
	count := 1
	power := radix
	for n >= power {
		n -= power
		count++
		power *= radix
	}
	result := []byte{}
	for ; count > 0; count-- {
		result = append(result, byte((n%radix)+65))
		n = int(math.Trunc(float64(n) / float64(radix)))
	}
	for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = result[j], result[i]
	}
	return string(result)
}

func AlphabetToNumber(alphabet string) int {
	if alphabet == "" {
		return -1
	}
	upper := strings.ToUpper(alphabet)
	radix := 26
	power := 1
	result := -1
	for i := len(upper) - 1; i >= 0; i-- {
		result += (int(upper[i])-65)*power + power
		power *= radix
	}
	return result
}
